### PR TITLE
add smooth_l1 op to ngraph emitter

### PR DESCRIPTION
**Pull request by ashokei**
## Description ##
add smooth_l1 op to ngraph emitter

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/ngraph/ngraph_emitter.cc [hunk 1]
```diff
@@ -695,6 +695,42 @@ void Emitter::CreateBinaryOps() {
...
   };
+  ngraph_op_funcs_["smooth_l1"] = [this](const NodePtr& node) {
+    /* Smooth L1 Loss is a loss specific for R-CNN franchise training
+     * Smooth L1 Loss function:
+     * f(x) = 0.5 * (sigma * x) ^ 2,     |x| < 1 / sigma^2
+     *      = |x| - 0.5 / sigma / sigma, otherwise
+     * When sigma = 1, it is equivalent to the Huber loss, evaluated at
+     * delta = 1.
+     * smooth_l1_loss = w_out * f(w_in * x)
+     * with w_in, w_out provided by input_data.
```
Comment (cconvey) [inline src/ngraph/ngraph_emitter.cc]: IMO the way this block-comment is structured is a bit confusing.  Lines 700-702 are worded in a way that sounds like you're giving the complete description of this operator's semantics.  But then lines 705-706 add more computation to the definition.

File: src/ngraph/ngraph_emitter.cc [hunk 2]
```diff
@@ -695,6 +695,42 @@ void Emitter::CreateBinaryOps() {
...
+     */
+    auto input = op_map_[node->inputs_[0]];
+    auto sigma =
+        makeConstant(node, get_default(node, "scalar", std::string("0")));
+    auto sigma_sq = sigma * sigma;
+    auto inv_sigma_sq = std::make_shared<ngraph::op::Divide>(
+        makeConstant(node, "1.0"), sigma_sq);
```
Comment (cconvey) [inline src/ngraph/ngraph_emitter.cc]: Can/should we sniff for the value of sigma being zero?

We *can* do that sniffing here because it's a graph-compile-time constant.

We *maybe* *should* do that sniffing here because it's often a good idea to let users know about problems ASAP, so they can better identify  what they're doing wrong on their end.

File: src/ngraph/ngraph_emitter.cc [hunk 3]
```diff
@@ -695,6 +695,42 @@ void Emitter::CreateBinaryOps() {
...
+
+    // check if input is greater than inv_sigma_sq
+    auto is_input_gt_sigma =
+        std::make_shared<ngraph::op::Greater>(input, inv_sigma_sq);
```
Comment (cconvey) [inline src/ngraph/ngraph_emitter.cc]: IMO this variable name is a big misleading, given that it's actually comparing the input to `inv_sigma_sq`.

File: src/ngraph/ngraph_emitter.cc [hunk 4]
```diff
@@ -695,6 +695,42 @@ void Emitter::CreateBinaryOps() {
...
+
+    // f(x) = 0.5 * (sigma * x) ^ 2
```
Comment (cconvey) [inline src/ngraph/ngraph_emitter.cc]: 💯 👍 🥇 for using comments to clarify what you intend to be computed by the nodes you're creating!

IMO it would be helpful for you to _not_ use the function name `f(x)` in this line and below, unless you're defining a function that specifically matches one of the two versions of `f(x)` that you describe in the block-comment at the top of this commit.

By using the same name `f(x)` in all of those places, it's unclear regarding which computations here and below are meant to exactly correspond fo the two definitions of `f(x)` you use in the block-comment above.

File: src/ngraph/ngraph_emitter.cc [hunk 5]
```diff
@@ -695,6 +695,42 @@ void Emitter::CreateBinaryOps() {
...
+    auto result_input_sq = makeConstant(node, "0.5") * input * input * sigma_sq;
+    // f(x) = x - 0.5 / sigma / sigma
+    auto result_input_gt = input - makeConstant(node, "0.5") * inv_sigma_sq;
+    // f(x) = -x - 0.5 / sigma / sigma
+    auto result_input_lt = -input - makeConstant(node, "0.5") * inv_sigma_sq;
```
Comment (cconvey) [inline src/ngraph/ngraph_emitter.cc]: Would it make sense to define just one nGraph node to represent the constant 0.5?  I don't know either way in terms of how that impacts nGraph's performance at graph-compile-time and graph-execution-time.

File: src/ngraph/ngraph_emitter.cc [hunk 6]
```diff
@@ -695,6 +695,42 @@ void Emitter::CreateBinaryOps() {
     return cast_result(CreateAutoBroadcast<ngraph::op::LessEq>(node),
                        getType(node->dtype_));
   };
+  ngraph_op_funcs_["smooth_l1"] = [this](const NodePtr& node) {
+    /* Smooth L1 Loss is a loss specific for R-CNN franchise training
+     * Smooth L1 Loss function:
+     * f(x) = 0.5 * (sigma * x) ^ 2,     |x| < 1 / sigma^2
+     *      = |x| - 0.5 / sigma / sigma, otherwise
```
[the remark this answers is not in the document]
Comment (ashokei) [reply src/ngraph/ngraph_emitter.cc]: will make the change, it will be consistent and more clear.

File: src/ngraph/ngraph_emitter.cc (hunk 1 above)
Comment (ashokei) [reply src/ngraph/ngraph_emitter.cc]: 705-706 is not formula, but related info. This comment block is copied from mxnet op description.

File: src/ngraph/ngraph_emitter.cc (hunk 2 above)
Comment (ashokei) [reply src/ngraph/ngraph_emitter.cc]: formula does not treat zero as special for sigma, it is same as any other `scalar` value the user chooses to pass. we should not restrict users from using this formula with zero (as formula or op description does not say so)

File: src/ngraph/ngraph_emitter.cc (hunk 3 above)
Comment (ashokei) [reply src/ngraph/ngraph_emitter.cc]: will change var name to inv_sq suffix

File: src/ngraph/ngraph_emitter.cc (hunk 5 above)
Comment (ashokei) [reply src/ngraph/ngraph_emitter.cc]: yep, will refactor that.

File: src/ngraph/ngraph_emitter.cc (hunk 2 above)
Comment (cconvey) [reply src/ngraph/ngraph_emitter.cc]: 👍 FYI, I created #253 for follow-up.
